File: src/main/java/com/zepben/blobstore/sqlite/SqliteConnectionFactory.java
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package com.zepben.blobstore.sqlite;

import com.zepben.annotations.EverythingIsNonnullByDefault;

import java.nio.file.Path;
import java.sql.*;
import java.util.Collections;
import java.util.HashSet;
import java.util.Set;

/**
 * ConnectionFactory that returns a new connection to an SQLite database every time getConnection() is called.
 */
@EverythingIsNonnullByDefault
public class SqliteConnectionFactory implements ConnectionFactory {

    static final String VERSION_TABLE = "schema_version";
    static final String SCHEMA_VERSION = "1";

    private final Set<String> tags;
    private final Path file;

    @SuppressWarnings("WeakerAccess")
    public SqliteConnectionFactory(Path file, Set<String> tags) {
        for (String tag : tags) {
            if (!supportedTableName(tag))
                throw new IllegalArgumentException("unsupported tag: " + tag);
        }

        this.tags = Set.copyOf(tags);
        this.file = file;
    }

    private boolean supportedTableName(String tag) {
        if (tag.equals(IdIndex.ID_INDEX_TABLE))
            return false;

        if (tag.equals(VERSION_TABLE))
            return false;

        for (int i = 0, n = tag.length(); i < n; ++i) {
            char c = tag.charAt(i);
            if (!(Character.isLetterOrDigit(c) || c == '_'))
                return false;
        }
        return true;
    }

    @Override
    public Connection getConnection() throws SQLException {
        Connection connection = null;
        try {
            connection = DriverManager.getConnection("jdbc:sqlite:file:" + file.toString() + "?cache=shared");

            checkVersion(connection);
            createTables(connection);

            return connection;
        } catch (SQLException e) {
            if (connection != null)
                connection.close();

            throw new SQLException(String.format("Failed to initialise sqlite database '%s'", file.toString()), e);
        }
    }

    private void checkVersion(Connection connection) throws SQLException {
        try (Statement stmt = connection.createStatement()) {
            try {
                String sql = String.format("SELECT name FROM sqlite_master WHERE type='table' AND name='%s'", VERSION_TABLE);
                boolean tableExists = stmt.executeQuery(sql).next();
                if (!tableExists) {
                    sql = String.format("CREATE TABLE %s (version TEXT)", VERSION_TABLE);
                    stmt.executeUpdate(sql);

                    sql = String.format("INSERT INTO %s (version) VALUES (%s)", VERSION_TABLE, SCHEMA_VERSION);
                    stmt.execute(sql);
                }
            } catch (SQLException e) {
                throw new SQLException("Failed to create version table", e);
            }

            String versionInDb = null;
            boolean matchingVersion = false;
            String sql = String.format("SELECT * FROM %s", VERSION_TABLE);
            ResultSet rs = stmt.executeQuery(sql);
            if (rs.next()) {
                versionInDb = rs.getString(1);
                matchingVersion = SCHEMA_VERSION.equals(versionInDb);
            }

            if (!matchingVersion) {
                String msg = String.format("Wrong version number, expected %s found %s", SCHEMA_VERSION, versionInDb);
                throw new SQLException(msg);
            }
        }
    }

    private void createTables(Connection connection) throws SQLException {
        try (Statement stmt = connection.createStatement()) {
            createIdIndexTable(stmt);
            createMetadataTable(stmt);
            createBlobTables(stmt);
        }
    }

    private void createIdIndexTable(Statement stmt) throws SQLException {
        String sql = String.format("CREATE TABLE IF NOT EXISTS %s (id INTEGER PRIMARY KEY AUTOINCREMENT, entity_id TEXT)", IdIndex.ID_INDEX_TABLE);
        try {
            stmt.execute(sql);
        } catch (SQLException e) {
            throw new SQLException("Failed to create id index table", e);
        }

        sql = String.format("CREATE UNIQUE INDEX IF NOT EXISTS %s_idx on %<s (entity_id)", IdIndex.ID_INDEX_TABLE);
        try {
            stmt.execute(sql);
        } catch (SQLException e) {
            throw new SQLException("Failed to create index on id index table", e);
        }
    }

    private void createMetadataTable(Statement stmt) throws SQLException {
        String sql = String.format("CREATE TABLE IF NOT EXISTS %s (key TEXT, value TEXT)", Metadata.TABLE_NAME);
        try {
            stmt.execute(sql);
        } catch (SQLException e) {
            throw new SQLException("Failed to create metadata table", e);
        }
    }

    private void createBlobTables(Statement stmt) throws SQLException {
        String sql = "CREATE TABLE IF NOT EXISTS %s (id INTEGER PRIMARY KEY, data BLOB) WITHOUT ROWID";
        try {
            for (String tag : tags) {
                stmt.execute(String.format(sql, tag));
            }
        } catch (SQLException e) {
            throw new SQLException("Failed to create blob tables", e);
        }
    }

}

File: src/main/java/com/zepben/blobstore/sqlite/SqliteConnectionFactory.kt
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package com.zepben.blobstore.sqlite

import java.nio.file.Path
import java.sql.*

/**
 * ConnectionFactory that returns a new connection to an SQLite database every time getConnection() is called.
 */
class SqliteConnectionFactory(
    private val file: Path,
    tags: Set<String>
) : ConnectionFactory {

    // Take a copy of the passed in tags.
    private val tags = tags.toSet().onEach { require(supportedTableName(it)) { "unsupported tag: $it" } }

    private fun supportedTableName(tag: String): Boolean =
        when (tag) {
            IdIndex.ID_INDEX_TABLE -> false
            VERSION_TABLE -> false
            else -> tag.none { (!(Character.isLetterOrDigit(it) || it == '_')) }
        }

    @Throws(SQLException::class)
    override fun getConnection(): Connection =
        try {
            DriverManager.getConnection("jdbc:sqlite:file:$file?cache=shared").apply {
                validateSchema()
            }
        } catch (e: SQLException) {
            throw SQLException("Failed to initialise sqlite database '$file'", e)
        }

    private fun Connection.validateSchema() {
        try {
            createStatement().use { stmt ->
                checkVersion(stmt)

                ensureIdIndexTable(stmt)
                ensureMetadataTable(stmt)
                ensureBlobTables(stmt)
            }
        } catch (e: SQLException) {
            close()
            throw SQLException("Sqlite database '$file' failed schema validation", e)
        }
    }

    private fun checkVersion(stmt: Statement) {
        try {
            stmt.executeQuery("SELECT name FROM sqlite_master WHERE type='table' AND name='$VERSION_TABLE'").use { rs ->
                val tableExists = rs.next()
                if (!tableExists) {
                    stmt.executeUpdate("CREATE TABLE $VERSION_TABLE (version TEXT)")
                    stmt.execute("INSERT INTO $VERSION_TABLE (version) VALUES ($SCHEMA_VERSION)")
                }
            }
        } catch (e: SQLException) {
            throw SQLException("Failed to create version table", e)
        }

        var versionInDb: String? = null
        var matchingVersion = false
        stmt.executeQuery("SELECT * FROM $VERSION_TABLE").use { rs ->
            if (rs.next()) {
                versionInDb = rs.getString(1)
                matchingVersion = SCHEMA_VERSION == versionInDb
            }
        }

        if (!matchingVersion)
            throw SQLException("Wrong version number, expected $SCHEMA_VERSION found $versionInDb")
    }

    private fun ensureIdIndexTable(stmt: Statement) {
        try {
            stmt.execute("CREATE TABLE IF NOT EXISTS ${IdIndex.ID_INDEX_TABLE} (id INTEGER PRIMARY KEY AUTOINCREMENT, entity_id TEXT)")
        } catch (e: SQLException) {
            throw SQLException("Failed to create id index table", e)
        }

        try {
            stmt.execute("CREATE UNIQUE INDEX IF NOT EXISTS ${IdIndex.ID_INDEX_TABLE}_idx on ${IdIndex.ID_INDEX_TABLE} (entity_id)")
        } catch (e: SQLException) {
            throw SQLException("Failed to create index on id index table", e)
        }
    }

    private fun ensureMetadataTable(stmt: Statement) {
        try {
            stmt.execute("CREATE TABLE IF NOT EXISTS ${Metadata.TABLE_NAME} (key TEXT, value TEXT)")
        } catch (e: SQLException) {
            throw SQLException("Failed to create metadata table", e)
        }
    }

    private fun ensureBlobTables(stmt: Statement) {
        try {
            for (tag in tags) {
                stmt.execute("CREATE TABLE IF NOT EXISTS $tag (id INTEGER PRIMARY KEY, data BLOB) WITHOUT ROWID")
            }
        } catch (e: SQLException) {
            throw SQLException("Failed to create blob tables", e)
        }
    }

    companion object {
        const val VERSION_TABLE = "schema_version"
        const val SCHEMA_VERSION = "1"
    }

}
